File: countdown.js
// countdown.js
const fs = require('fs');
const readline = require('readline');

const EVENTS_FILE = 'events.json';
let events = [];

function loadEvents() {
    try {
        const data = fs.readFileSync(EVENTS_FILE, 'utf8');
        events = JSON.parse(data);
    } catch (e) {
        events = [];
    }
}

function saveEvents() {
    fs.writeFileSync(EVENTS_FILE, JSON.stringify(events, null, 2));
}

function parseDate(dateStr) {
    const formats = [
        /^(\d{4})-(\d{2})-(\d{2})$/,
        /^(\d{2})\/(\d{2})\/(\d{4})$/,
        /^(\d{2})\/(\d{2})\/(\d{4})$/,
        /^(\d{2})\.(\d{2})\.(\d{4})$/
    ];
    for (const fmt of formats) {
        const match = dateStr.match(fmt);
        if (match) {
            let y, m, d;
            if (fmt === formats[0]) { y = match[1]; m = match[2]; d = match[3]; }
            else if (fmt === formats[1]) { d = match[1]; m = match[2]; y = match[3]; }
            else if (fmt === formats[2]) { m = match[1]; d = match[2]; y = match[3]; }
            else { d = match[1]; m = match[2]; y = match[3]; }
            const date = new Date(parseInt(y), parseInt(m)-1, parseInt(d));
            if (!isNaN(date)) return date;
        }
    }
    return null;
}

function findEvent(name) {
    return events.find(e => e.name.toLowerCase() === name.toLowerCase());
}

function addEvent(name, dateStr, timeStr) {
    const date = parseDate(dateStr);
    if (!date) return false;
    if (!timeStr) timeStr = '00:00';
    const [h, m] = timeStr.split(':').map(Number);
    date.setHours(h || 0, m || 0, 0, 0);
    events.push({ name, date: dateStr, time: timeStr });
    saveEvents();
    return true;
}

function deleteEvent(name) {
    const idx = events.findIndex(e => e.name.toLowerCase() === name.toLowerCase());
    if (idx !== -1) {
        events.splice(idx, 1);
        saveEvents();
        return true;
    }
    return false;
}

function formatCountdown(ms) {
    if (ms < 0) return 'Event has passed!';
    const totalSec = Math.floor(ms / 1000);
    const years = Math.floor(totalSec / (365.25 * 24 * 3600));
    let rem = totalSec - years * (365.25 * 24 * 3600);
    const months = Math.floor(rem / (30.44 * 24 * 3600));
    rem -= months * (30.44 * 24 * 3600);
    const days = Math.floor(rem / (24 * 3600));
    rem -= days * (24 * 3600);
    const hours = Math.floor(rem / 3600);
    rem -= hours * 3600;
    const minutes = Math.floor(rem / 60);
    const seconds = rem % 60;
    const parts = [];
    if (years > 0) parts.push(`${years} year${years > 1 ? 's' : ''}`);
    if (months > 0) parts.push(`${months} month${months > 1 ? 's' : ''}`);
    if (days > 0) parts.push(`${days} day${days > 1 ? 's' : ''}`);
    if (hours > 0) parts.push(`${hours} hour${hours > 1 ? 's' : ''}`);
    if (minutes > 0) parts.push(`${minutes} minute${minutes > 1 ? 's' : ''}`);
    if (seconds > 0) parts.push(`${seconds} second${seconds > 1 ? 's' : ''}`);
    return parts.length ? parts.join(', ') : '0 seconds';
}

function countdown(event, formatStyle, live) {
    const [h, m] = (event.time || '00:00').split(':').map(Number);
    const target = new Date(event.date + 'T' + String(h||0).padStart(2,'0') + ':' + String(m||0).padStart(2,'0'));
    const now = new Date();
    let diff = target - now;
    if (diff <= 0) {
        console.log(`⏰ ${event.name} is in the past!`);
        return;
    }
    if (live) {
        const interval = setInterval(() => {
            const now2 = new Date();
            const diff2 = target - now2;
            if (diff2 <= 0) {
                clearInterval(interval);
                console.log(`\r🎉 ${event.name} has arrived!    `);
                return;
            }
            process.stdout.write(`\r⏳ ${event.name}: ${formatCountdown(diff2)}    `);
        }, 1000);
        setTimeout(() => {}, 10);
    } else {
        if (formatStyle === 'compact') {
            const totalSec = Math.floor(diff / 1000);
            const days = Math.floor(totalSec / 86400);
            const hours = Math.floor((totalSec % 86400) / 3600);
            const minutes = Math.floor((totalSec % 3600) / 60);
            const seconds = totalSec % 60;
            console.log(`⏳ ${event.name}: ${String(days).padStart(2,'0')}:${String(hours).padStart(2,'0')}:${String(minutes).padStart(2,'0')}:${String(seconds).padStart(2,'0')}`);
        } else if (formatStyle === 'short') {
            console.log(`⏳ ${event.name}: ${Math.floor(diff / 1000)} seconds`);
        } else {
            console.log(`⏳ ${event.name}: ${formatCountdown(diff)}`);
        }
    }
}

function listEvents() {
    if (events.length === 0) {
        console.log('No events saved.');
        return;
    }
    for (const e of events) {
        const [h, m] = (e.time || '00:00').split(':').map(Number);
        const target = new Date(e.date + 'T' + String(h||0).padStart(2,'0') + ':' + String(m||0).padStart(2,'0'));
        const status = target > new Date() ? '⏳ pending' : '✅ passed';
        console.log(`  ${e.name}: ${e.date} ${e.time} - ${status}`);
    }
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function ask(question) {
    return new Promise(resolve => rl.question(question, resolve));
}

async function interactive() {
    loadEvents();
    console.log('=== Countdown Timer ===');
    while (true) {
        console.log('\n1. Set new event');
        console.log('2. Show countdown');
        console.log('3. Live countdown');
        console.log('4. List events');
        console.log('5. Delete event');
        console.log('6. Save events');
        console.log('7. Load events');
        console.log('8. Exit');
        const choice = await ask('Choose: ');
        switch (choice.trim()) {
            case '1': {
                const name = await ask('Event name: ');
                if (!name.trim()) { console.log('Name required.'); break; }
                if (findEvent(name.trim())) { console.log('Event already exists.'); break; }
                const dateStr = await ask('Event date (YYYY-MM-DD): ');
                if (!parseDate(dateStr.trim())) { console.log('Invalid date.'); break; }
                let timeStr = await ask('Event time (HH:MM, optional): ');
                if (!timeStr.trim()) timeStr = '00:00';
                if (addEvent(name.trim(), dateStr.trim(), timeStr.trim())) {
                    console.log('Event saved.');
                }
                break;
            }
            case '2': {
                const name = await ask('Event name: ');
                const event = findEvent(name.trim());
                if (!event) { console.log('Event not found.'); break; }
                countdown(event, 'full', false);
                break;
            }
            case '3': {
                const name = await ask('Event name: ');
                const event = findEvent(name.trim());
                if (!event) { console.log('Event not found.'); break; }
                countdown(event, 'full', true);
                break;
            }
            case '4': listEvents(); break;
            case '5': {
                const name = await ask('Event name to delete: ');
                if (deleteEvent(name.trim())) console.log('Event deleted.');
                else console.log('Event not found.');
                break;
            }
            case '6': saveEvents(); console.log('Events saved.'); break;
            case '7': loadEvents(); console.log('Events loaded.'); break;
            case '8': console.log('Goodbye!'); rl.close(); return;
            default: console.log('Invalid choice.');
        }
    }
}

function cli() {
    const args = require('minimist')(process.argv.slice(2));
    loadEvents();
    if (args.list) { listEvents(); return; }
    if (args.delete) {
        if (deleteEvent(args.delete)) console.log('Event deleted.');
        else console.log('Event not found.');
        return;
    }
    if (args.event && args.date) {
        const event = findEvent(args.event);
        if (event) {
            const formatStyle = args.format || 'full';
            const live = args.live || false;
            countdown(event, formatStyle, live);
        } else {
            console.log('Event not found.');
        }
    } else if (args.event) {
        const event = findEvent(args.event);
        if (event) countdown(event, args.format || 'full', args.live || false);
        else console.log('Event not found.');
    } else {
        console.log('Usage: node countdown.js --event NAME --date DATE [--time TIME] [--format full|compact|short] [--live]');
    }
}

if (process.argv.length > 2) {
    cli();
} else {
    interactive();
}

File: countdown.rb
# countdown.rb
require 'json'
require 'date'
require 'optparse'
require 'time'

EVENTS_FILE = 'events.json'
$events = []

def load_events
  $events = JSON.parse(File.read(EVENTS_FILE)) rescue []
end

def save_events
  File.write(EVENTS_FILE, JSON.pretty_generate($events))
end

def parse_date(date_str)
  formats = ['%Y-%m-%d', '%d/%m/%Y', '%m/%d/%Y', '%d.%m.%Y']
  formats.each do |fmt|
    begin
      return Date.strptime(date_str, fmt)
    rescue ArgumentError
    end
  end
  nil
end

def find_event(name)
  $events.find { |e| e['name'].downcase == name.downcase }
end

def add_event(name, date_str, time_str)
  return false unless parse_date(date_str)
  time_str = '00:00' if time_str.nil? || time_str.empty?
  $events << { 'name' => name, 'date' => date_str, 'time' => time_str }
  save_events
  true
end

def delete_event(name)
  idx = $events.index { |e| e['name'].downcase == name.downcase }
  if idx
    $events.delete_at(idx)
    save_events
    true
  else
    false
  end
end

def format_countdown(seconds)
  return 'Event has passed!' if seconds < 0
  years = (seconds / (365.25 * 24 * 3600)).to_i
  seconds -= (years * 365.25 * 24 * 3600).to_i
  months = (seconds / (30.44 * 24 * 3600)).to_i
  seconds -= (months * 30.44 * 24 * 3600).to_i
  days = (seconds / (24 * 3600)).to_i
  seconds -= days * 24 * 3600
  hours = (seconds / 3600).to_i
  seconds -= hours * 3600
  minutes = (seconds / 60).to_i
  seconds = seconds % 60
  parts = []
  parts << "#{years} year#{years > 1 ? 's' : ''}" if years > 0
  parts << "#{months} month#{months > 1 ? 's' : ''}" if months > 0
  parts << "#{days} day#{days > 1 ? 's' : ''}" if days > 0
  parts << "#{hours} hour#{hours > 1 ? 's' : ''}" if hours > 0
  parts << "#{minutes} minute#{minutes > 1 ? 's' : ''}" if minutes > 0
  parts << "#{seconds} second#{seconds > 1 ? 's' : ''}" if seconds > 0
  parts.empty? ? '0 seconds' : parts.join(', ')
end

def countdown(event, format_style, live)
  target = Time.parse("#{event['date']} #{event['time']}")
  now = Time.now
  diff = target - now
  if diff <= 0
    puts "⏰ #{event['name']} is in the past!"
    return
  end
  if live
    while diff > 0
      print "\r⏳ #{event['name']}: #{format_countdown(diff)}    "
      sleep 1
      now = Time.now
      diff = target - now
    end
    puts "\r🎉 #{event['name']} has arrived!    "
  else
    if format_style == 'compact'
      total_sec = diff.to_i
      days = total_sec / 86400
      hours = (total_sec % 86400) / 3600
      minutes = (total_sec % 3600) / 60
      seconds = total_sec % 60
      puts "⏳ #{event['name']}: #{'%02d' % days}:#{'%02d' % hours}:#{'%02d' % minutes}:#{'%02d' % seconds}"
    elsif format_style == 'short'
      puts "⏳ #{event['name']}: #{diff.to_i} seconds"
    else
      puts "⏳ #{event['name']}: #{format_countdown(diff)}"
    end
  end
end

def list_events
  if $events.empty?
    puts 'No events saved.'
    return
  end
  $events.each do |e|
    target = Time.parse("#{e['date']} #{e['time']}")
    status = target > Time.now ? '⏳ pending' : '✅ passed'
    puts "  #{e['name']}: #{e['date']} #{e['time']} - #{status}"
  end
end

def interactive
  load_events
  puts "=== Countdown Timer ==="
  loop do
    puts "\n1. Set new event"
    puts "2. Show countdown"
    puts "3. Live countdown"
    puts "4. List events"
    puts "5. Delete event"
    puts "6. Save events"
    puts "7. Load events"
    puts "8. Exit"
    print "Choose: "
    choice = gets.chomp.strip
    case choice
    when '1'
      print "Event name: "
      name = gets.chomp.strip
      if name.empty?
        puts "Name required."
        next
      end
      if find_event(name)
        puts "Event already exists."
        next
      end
      print "Event date (YYYY-MM-DD): "
      date_str = gets.chomp.strip
      unless parse_date(date_str)
        puts "Invalid date."
        next
      end
      print "Event time (HH:MM, optional): "
      time_str = gets.chomp.strip
      if add_event(name, date_str, time_str)
        puts "Event saved."
      end
    when '2'
      print "Event name: "
      name = gets.chomp.strip
      event = find_event(name)
      if event
        countdown(event, 'full', false)
      else
        puts "Event not found."
      end
    when '3'
      print "Event name: "
      name = gets.chomp.strip
      event = find_event(name)
      if event
        countdown(event, 'full', true)
      else
        puts "Event not found."
      end
    when '4'
      list_events
    when '5'
      print "Event name to delete: "
      name = gets.chomp.strip
      if delete_event(name)
        puts "Event deleted."
      else
        puts "Event not found."
      end
    when '6'
      save_events
      puts "Events saved."
    when '7'
      load_events
      puts "Events loaded."
    when '8'
      puts "Goodbye!"
      break
    else
      puts "Invalid choice."
    end
  end
end

def cli
  options = {}
  OptionParser.new do |opts|
    opts.on('--event NAME', 'Event name') { |v| options[:event] = v }
    opts.on('--date DATE', 'Event date') { |v| options[:date] = v }
    opts.on('--time TIME', 'Event time') { |v| options[:time] = v }
    opts.on('--format FORMAT', 'Output format') { |v| options[:format] = v }
    opts.on('--live', 'Live countdown') { options[:live] = true }
    opts.on('--list', 'List events') { options[:list] = true }
    opts.on('--delete NAME', 'Delete event') { |v| options[:delete] = v }
  end.parse!
  load_events
  if options[:list]
    list_events
    return
  end
  if options[:delete]
    if delete_event(options[:delete])
      puts "Event deleted."
    else
      puts "Event not found."
    end
    return
  end
  if options[:event] && options[:date]
    event = find_event(options[:event])
    if event
      countdown(event, options[:format] || 'full', options[:live] || false)
    else
      puts "Event not found."
    end
  else
    puts "Usage: ruby countdown.rb --event NAME --date DATE [--time TIME] [--format full|compact|short] [--live]"
  end
end

if ARGV.empty?
  interactive
else
  cli
end
